app/src/main/java/ai/elimu/appstore/synchronization/AppListArrayAdapter.java

```diff
@@ -66,12 +94,179 @@ public View getView(int position, @Nullable View convertView, @NonNull ViewGroup
+
+        private ApplicationVersionGson applicationVersionGson;
+
+        private ProgressBar progressBarDownloadProgress;
+        private TextView textViewDownloadProgress;
+
+        public DownloadApplicationAsyncTask(ProgressBar progressBarDownloadProgress, TextView textViewDownloadProgress) {
+            this.progressBarDownloadProgress = progressBarDownloadProgress;
+            this.textViewDownloadProgress = textViewDownloadProgress;
+        }
+
+        @Override
+        protected Integer doInBackground(ApplicationVersionGson... applicationVersionGsons) {
+            Timber.i("doInBackground");
+
+            applicationVersionGson = applicationVersionGsons[0];
+            Timber.i("applicationVersionGson.getFileUrl(): " + applicationVersionGson.getFileUrl());
+            Timber.i("applicationVersionGson.getFileSizeInKb(): " + applicationVersionGson.getFileSizeInKb());
+
+            // Reset to initial state
+            Integer fileSizeInKbsDownloaded = 0;
+            publishProgress(fileSizeInKbsDownloaded);
+
+            // Download APK file and store it on SD card
+            String fileUrl = BuildConfig.BASE_URL + applicationVersionGson.getFileUrl() +
+                    "?deviceId=" + DeviceInfoHelper.getDeviceId(context) +
+                    "&checksum=" + ChecksumHelper.getChecksum(context) +
+                    "&locale=" + UserPrefsHelper.getLocale(context) +
+                    "&deviceModel=" + DeviceInfoHelper.getDeviceModel(context) +
+                    "&osVersion=" + Build.VERSION.SDK_INT +
+                    "&applicationId=" + DeviceInfoHelper.getApplicationId(context) +
+                    "&appVersionCode=" + DeviceInfoHelper.getAppVersionCode(context);
+            Timber.i("fileUrl: " + fileUrl);
+
+            String fileName = applicationVersionGson.getApplication().getPackageName() + "-" + applicationVersionGson.getVersionCode() + ".apk";
+            Timber.i("fileName: " + fileName);
+
+            Timber.i("Downloading APK: " + applicationVersionGson.getApplication().getPackageName() + " (version " + applicationVersionGson.getVersionCode() + ", " + applicationVersionGson.getFileSizeInKb() + "kB)");
+
+//            File apkFile = ApkLoader.loadApk(fileUrl, fileName, context);
+            String urlValue = fileUrl;
+            // Copied from ApkLoader:
+            Timber.i("Downloading from " + urlValue + "...");
+
+            String language = Locale.getDefault().getLanguage();
+            File apkDirectory = new File(Environment.getExternalStorageDirectory() + "/.elimu-ai/appstore/apks/" + language);
+            Timber.i("apkDirectory: " + apkDirectory);
+            if (!apkDirectory.exists()) {
+                apkDirectory.mkdirs();
+            }
+
+            File apkFile = new File(apkDirectory, fileName);
+            Timber.i("apkFile: " + apkFile);
+            Timber.i("apkFile.exists(): " + apkFile.exists());
+
+            if (!apkFile.exists()) {
```

Comment:
I would try to reuse a download method/function, since you already have one for content download

Reply:
Yes, I agree. I'll plan in some refactoring of this code.